### PR TITLE
cleanup: use kubebuilder marker instead of webhook

### LINKED ISSUE #15: Use kubebuilder validation & default for simplification
No longer to add specific logics to webhook, See https://book.kubebuilder.io/reference/markers/crd-validation
/kind feature

### PR DESCRIPTION
**Pull request by nayihz**
#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes

Fixes #15 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/webhooks/leaderworkerset_webhook.go [hunk 1]
```diff
@@ -51,14 +51,6 @@ var _ webhook.CustomDefaulter = &LeaderWorkerSetWebhook{}
 func (r *LeaderWorkerSetWebhook) Default(ctx context.Context, obj runtime.Object) error {
 	lws := obj.(*v1.LeaderWorkerSet)
 
-	if lws.Spec.LeaderWorkerTemplate.RestartPolicy == "" {
```
Comment (kerthcet) [inline pkg/webhooks/leaderworkerset_webhook.go:54]: Default only works for optional field, plz revert this change.

File: api/leaderworkerset/v1/leaderworkerset_types.go [hunk 2]
```diff
@@ -110,6 +110,8 @@ type LeaderWorkerTemplate struct {
 	Size *int32 `json:"size,omitempty"`
 
 	// RestartPolicy defines the restart policy when pod failures happen.
+	// +kubebuilder:default=Default
+	// +kubebuilder:validation:Enum=Default;RecreateGroupOnPodRestart
```
Comment (kerthcet) [inline api/leaderworkerset/v1/leaderworkerset_types.go:114]:
```suggestion
	// +kubebuilder:validation:Enum={Default,RecreateGroupOnPodRestart}
```

**Review by kerthcet — COMMENTED** [review]
Also we forgot to add a validation for maxUnavailable, which can be less t than 1. Can you help to fix that? Don't forget to add a test.

Thanks.

File: pkg/webhooks/leaderworkerset_webhook.go (hunk 1 above)
Comment (nayihz) [reply pkg/webhooks/leaderworkerset_webhook.go:54]: `// +kubebuilder:default=Default` also works for `RestartPolicy`

**Comment by nayihz** [comment]
> Also we forgot to add a validation for maxUnavailable, which can be less t than 1.

`maxUnavailable` could be int or string. IIRC we have no marker to validate it. Maybe we can only validate this field in the webhook?

File: test/integration/webhooks/leaderworkerset_test.go [hunk 3]
```diff
@@ -111,10 +111,15 @@ var _ = ginkgo.Describe("leaderworkerset defaulting, creation and update", func(
 		}),
 		ginkgo.Entry("apply default rollout strategy", &testDefaultingCase{
 			makeLeaderWorkerSet: func(ns *corev1.Namespace) *testutils.LeaderWorkerSetWrapper {
-				return testutils.BuildLeaderWorkerSet(ns.Name).RolloutStrategy(leaderworkerset.RolloutStrategy{}) // unset rollout strategy
```
Comment (kerthcet) [inline test/integration/webhooks/leaderworkerset_test.go:114]: Why change this? BuilderLeaderWorkerSet also set the rolloutStrategy, so what you changed here will not validate the defaulter for rollout strategy.

File: pkg/webhooks/leaderworkerset_webhook.go (hunk 1 above)
Comment (kerthcet) [reply pkg/webhooks/leaderworkerset_webhook.go:54]: You can, but it doesn't work. Default works after admission of the request, so `Required` will check first.

File: pkg/webhooks/leaderworkerset_webhook.go [hunk 4]
```diff
@@ -108,13 +100,10 @@ func (r *LeaderWorkerSetWebhook) generalValidate(ctx context.Context, obj runtim
...
 	}
-	if lws.Spec.LeaderWorkerTemplate.RestartPolicy != v1.DefaultRestartPolicy && lws.Spec.LeaderWorkerTemplate.RestartPolicy != v1.RecreateGroupOnPodRestart {
-		allErrs = append(allErrs, field.Invalid(specPath.Child("leaderWorkerTemplate", "restartPolicy"), lws.Spec.LeaderWorkerTemplate.RestartPolicy, "only support [Default, RecreateGroupOnPodRestart]"))
 
-	}
 	if lws.Spec.RolloutStrategy.RollingUpdateConfiguration != nil {
 		value, err := intstr.GetScaledValueFromIntOrPercent(&lws.Spec.RolloutStrategy.RollingUpdateConfiguration.MaxUnavailable, int(*lws.Spec.Replicas), false)
-		if err != nil || value > int(*lws.Spec.Replicas) {
+		if err != nil || value > int(*lws.Spec.Replicas) || value < 1 {
 			allErrs = append(allErrs, field.Invalid(specPath.Child("rolloutStrategy", "rollingUpdateConfiguration", "maxUnavailable"), lws.Spec.RolloutStrategy.RollingUpdateConfiguration.MaxUnavailable, "maxUnavailable must not greater than replicas"))
```
Comment (kerthcet) [inline pkg/webhooks/leaderworkerset_webhook.go:107]:
```suggestion
allErrs = append(allErrs, field.Invalid(specPath.Child("rolloutStrategy", "rollingUpdateConfiguration", "maxUnavailable"), lws.Spec.RolloutStrategy.RollingUpdateConfiguration.MaxUnavailable, "maxUnavailable can neither greater than replicas nor less than 1"))
```

File: pkg/webhooks/pod_webhook_test.go [hunk 5]
```diff
@@ -23,7 +23,6 @@ import (
 
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (kerthcet) [inline pkg/webhooks/pod_webhook_test.go:26]: good catch

File: pkg/webhooks/leaderworkerset_webhook.go (hunk 1 above)
Comment (kerthcet) [reply pkg/webhooks/leaderworkerset_webhook.go:54]: You successful here is because you set the `emitempty` to `Type` then this field is a **Optional** field now, but actually we want to make it a **Required** field. 

The problem with your approach here is the `RollingUpdateConfiguration` might be nil and panic when rolling update because `RollingUpdateConfiguration` can't be nil, but you didn't validate this below when spec.rolloutStrategy.type is empty.

My suggestion is:
- Remove the `omitempty` from `RestartPolicy` and `Type` to make them required fields
- Revert the codes here

File: pkg/webhooks/leaderworkerset_webhook.go [hunk 6]
```diff
@@ -51,14 +51,6 @@ var _ webhook.CustomDefaulter = &LeaderWorkerSetWebhook{}
...
 
-	if lws.Spec.LeaderWorkerTemplate.RestartPolicy == "" {
-		lws.Spec.LeaderWorkerTemplate.RestartPolicy = v1.DefaultRestartPolicy
-	}
-
-	if lws.Spec.RolloutStrategy.Type == "" {
-		lws.Spec.RolloutStrategy.Type = v1.RollingUpdateStrategyType
-	}
-
 	if lws.Spec.RolloutStrategy.Type == v1.RollingUpdateStrategyType && lws.Spec.RolloutStrategy.RollingUpdateConfiguration == nil {
```
Comment (kerthcet) [inline pkg/webhooks/leaderworkerset_webhook.go:54]: I mean here lws.Spec.RolloutStrategy.Type might be empty if we didn't set the value.

File: pkg/webhooks/leaderworkerset_webhook.go (hunk 1 above)
Comment (kerthcet) [reply pkg/webhooks/leaderworkerset_webhook.go:54]: See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

Comment (nayihz) [reply pkg/webhooks/leaderworkerset_webhook.go:54]: Thanks for detailed explanation.

**Review by kerthcet — COMMENTED** [review]
/lgtm
/approve

Thanks!

File: pkg/webhooks/leaderworkerset_webhook.go (hunk 1 above)
Comment (nayihz) [reply pkg/webhooks/leaderworkerset_webhook.go:54]: I'm curious that why users can set `RestartPolicy == ""` if `RestartPolicy` field is `required`

Comment (nayihz) [reply pkg/webhooks/leaderworkerset_webhook.go:54]: > RestartPolicy is required by the Schema because it can't be empty.

Since RestartPolicy cannot be configured as empty, why is it still possible to be set to`""`(`RestartPolicy == ""`)?

Comment (nayihz) [reply pkg/webhooks/leaderworkerset_webhook.go:54]: IIUC, RestartPolicy field is `required` and could be only set to `Default` or `RecreateGroupOnPodRestart`. So we should not allow user to set `RestartPolicy = ""`.
